### DISCUSSION TITLE
Altering Text Properties in a PDF

### DISCUSSION THREAD

**Discussion by Aadil-5122** [pymupdf/PyMuPDF#2222] [Looking for help]
Good Morning sir, 

Could you please suggest an efficient way to alter Text Properties of text in a pdf. 

So, what i want is to change:

**1. Size of Text
2. Bold/ Un-Bold Text
3. Text Offset**
(of particular text in a PDF)

What i thought was a very in-efficient approach of:
For Text Offset
-> first extracting text box and then deleting the text on the PDF,
then pasting the extracted to new co-ordinates.

For bold
-> I cant still figure out

For Size
-> first extracting text box and then deleting the text on the PDF,
then pasting the extracted (with new size) again.

Any scope to increase efficiency?

Thanks.

**Comment by JorjMcKie** [comment]
Text cannot be altered "in-place" - it must always be rewritten. E.g. there is no "switching" between bold / unbold, because these are different fonts!
Of course this is true for your other two use cases anyway: size of text, position change.

There **_might_** be a change to have the current font xref be replaced by another one pointing to a font with a different weight or style. But this (1) requires low-level code, and (2) automatically applies to all text in the whole document that uses the old font, (3) for some fonts, italic / bold / regular versions also have different character distances, so the result of such a change may very well look awkward.

**Reply by Aadil-5122** [reply]
Alright Sir, Leaving Weight and Style of Text, what is the most efficient approach to achieve the other two motives (Text Size and the Position)?

**Reply by JorjMcKie** [reply]
> what is the most efficient approach to achieve the other two motives (Text Size and the Position)?

0. get hold of the font used by the text in question
1. Remove the text using redaction annotations
2. re-insert the text with desired new font size and location

To find the font used, do `page.get_text("dict", ...)` **after** you have set a global parameter like this: `fitz.TOOLS.set_subset_fontnames(True)`. This will make sure that the span dictionaries contain the full font names, including any subset prefixes.
When iterating over the dictionary hierarchy of `page.get_text("dict")` look at the span containing your desired text. When encountering it, look at `span["font"]` and find its reference name in `page.get_fonts()` - using the fontname.

So you must 
1. first extract the text, then locate the font(s) of text(s) to reposition / resize.
2. Then remove text you want to change (redaction annotations).
3. Insert with new size / position, specifying the reference name determined before.
4.

**Reply by Aadil-5122** [reply]
"reference name" of? the font?

**Reply by JorjMcKie** [reply]
Ever looked at `page.get_fonts()`? These are F30 and F32 in the following. 'TLUEMO+TeXGyreHeros-Bold' is what you will find in text extraction.
```python
In [1]: import fitz
In [2]: doc=fitz.open("pymupdf.pdf")
In [3]: page=doc[0]
In [4]: page.get_fonts()
Out[4]:
[(1123, 'pfa', 'Type1', 'TLUEMO+TeXGyreHeros-Bold', 'F30', ''),
 (1124, 'pfa', 'Type1', 'UJFJBV+TeXGyreHeros-BoldItalic', 'F32', '')]
```
To use these fonts in text re-insertion of font span:
```python
page.insert_text(newstartpoint, span["text"], fontsize=newsize, fontname="F30"))
```

**Reply by Aadil-5122** [reply]
Alright Sir, Thanks Alot!
